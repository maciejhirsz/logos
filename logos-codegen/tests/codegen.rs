use insta::assert_snapshot;
use std::{error::Error, path::PathBuf};
use syn::parse_file;

#[rstest::rstest]
#[case("simple")]
#[case("no_error_lut")]
#[case("skip_callback")]
#[case("skip_callback_failure")]
#[case("error_callback0")]
#[case("error_callback1")]
#[case("error_callback_failure")]
#[case("prio_conflict")]
#[case("illegal_utf8")]
pub fn test_codegen(#[case] fixture: &str) -> Result<(), Box<dyn Error>> {
    let rust_ver = if cfg!(rust_1_82) { "1_82" } else { "pre_1_82" };

    let codegen_alg = if cfg!(feature = "state_machine_codegen") {
        "state_machine"
    } else {
        "tailcall"
    };

    let input_dir_path = [env!("CARGO_MANIFEST_DIR"), "tests", "data", "codegen"]
        .iter()
        .collect::<PathBuf>();

    let input_path = input_dir_path.join(format!("{}.rs", fixture));
    let input = std::fs::read_to_string(input_path)?;

    let generated = logos_codegen::generate(input.parse()?).to_string();

    let formatted =
        prettyplease::unparse(&parse_file(&generated).expect("Logos output is unparseable"));

    assert_snapshot!(format!("{fixture}-{codegen_alg}-{rust_ver}"), formatted);

    Ok(())
}
