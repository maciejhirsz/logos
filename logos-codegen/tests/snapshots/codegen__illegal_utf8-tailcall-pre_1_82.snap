---
source: logos-codegen/tests/codegen.rs
expression: formatted
---
impl<'s> ::logos::Logos<'s> for Token {
    type Error = ();
    type Extras = ();
    type Source = str;
    fn lex(
        lex: &mut ::logos::Lexer<'s, Self>,
    ) -> std::option::Option<std::result::Result<Self, Self::Error>> {
        fn _logos_derive_compile_errors() {
            {
                compile_error!(
                    "UTF-8 mode is requested, but the subpattern stuff = b\"\\\\xFF\" can match invalid utf8.\nYou can disable UTF-8 mode with #[logos(utf8 = false)]"
                )
            }
            {
                compile_error!(
                    "UTF-8 mode is requested, but the pattern b\"b\\xFF\" of variant `::B` can match invalid utf8.\nYou can disable UTF-8 mode with #[logos(utf8 = false)]"
                )
            }
            {
                compile_error!(
                    "UTF-8 mode is requested, but the pattern \"(?-u)2\\\\xFF\" of variant `::Two` can match invalid utf8.\nYou can disable UTF-8 mode with #[logos(utf8 = false)]"
                )
            }
            {
                compile_error!(
                    "UTF-8 mode is requested, but the pattern b\"3\\\\xFF\" of variant `::Three` can match invalid utf8.\nYou can disable UTF-8 mode with #[logos(utf8 = false)]"
                )
            }
        }
        unimplemented!()
    }
}
